internal/datastore/crdb/tx.go

```diff
@@ -37,36 +50,81 @@ type transactionFn func(tx pgx.Tx) error
 
 type executeTxRetryFunc func(context.Context, conn, pgx.TxOptions, transactionFn) error
 
+// RetryError wraps an error that prevented the transaction function from being retried.
+type RetryError struct {
+	err error
+}
+
+// Error returns the wrapped error
+func (e RetryError) Error() string {
+	return fmt.Errorf("unable to retry conflicted transaction: %w", e.err).Error()
+}
+
+// Unwrap returns the wrapped, non-retriable error
+func (e RetryError) Unwrap() error {
+	return e.err
+}
+
 func executeWithMaxRetries(max int) executeTxRetryFunc {
 	return func(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn) (err error) {
-		return execute(ctx, conn, txOptions, fn, max)
+		return executeWithResets(ctx, conn, txOptions, fn, max)
 	}
 }
 
-// adapted from https://github.com/cockroachdb/cockroach-go
-func execute(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn, maxRetries int) (err error) {
+// executeWithResets executes transactionFn and resets the tx when ambiguous crdb errors are encountered.
+func executeWithResets(ctx context.Context, conn conn, txOptions pgx.TxOptions, fn transactionFn, maxRetries int) (err error) {
+	var retries, resets int
+	defer func() {
+		retryHistogram.Observe(float64(retries))
+		resetHistogram.Observe(float64(resets))
+	}()
+
 	var tx pgx.Tx
-	tx, err = conn.BeginTx(ctx, txOptions)
-	if err != nil {
-		return err
-	}
 	defer func() {
 		if err == nil {
-			_ = tx.Commit(ctx)
-			return
+			commitErr := tx.Commit(ctx)
```

Comment:
since the `BeginTx` has been moved after this, `tx` can be nil here now (it would be unlikely; every reset would have to fail, but it would be possible)